README.md
# Sitemap-generator.sh

Sitemap xml generator in posix shell.

## Description

This script crawls a web site from a given starting local URL and generates a Sitemap file in the format that is accepted by Google.
It does not follow links to other web sites or parent directory.

## Usage

  Usage:

     $ sitemap-generator.sh [-r|--remote <url>] [-l|--locale <url>] [-p|--priority <number>] [-f|--frequency <string>] [-i|--index <string>] [-d|--docroot <path>] [-A|--accept <list>] [-R|-reject <list>] [-o|--output-file] [-6] [-h|--help] [-v|--version] [-vv|--verbose|--debug]

  Example:

     $ sitemap-generator.sh -l https://localhost/foobar/ -r https://example.com -d /home/html/foobar -p 0.8 -f daily

  Options:

    -r|--remote <url>           Set the remote URL.

    -l|--local <url>            Set the local URL (ex. http://localhost/foobar/ )
                                Not with filename (ex. http://localhost/foo/bar.php)

    -p|--priority <value>       Set the priority. Valid values range from 0.0 to 1.0.
                                Default is "0.5".

    -f|--frequency <value>      Set the frequency. Valid values are:
                                always, hourly, daily, weekly, monthly, yearly, never
                                Default is "weekly".

    -i|--index <filename>       Set the name of index file.
                                The default filename is "index.php".

    -d|--docroot <path>         Set dhe "Doc Root".

    -A|--accept <list>          Comma-separated list of accepted extensions.
                                Default is all.

    -R|--reject <list>          Comma-separated list of rejected extensions.
                                Default is nothing.

    -o|--output-file <filename> Set the name of the geneated sitemap file.
                                The default file name is sitemap.xml.

    -6                          Set the inet6-only to wget.
                                Connect only to IPv6 addresses.

    -ssl|--check-ssl            Check if there are duplicate URLs with http and https.

    -vv|--verbose               Print details when crawling with wget.

    --debug                     Set bash to debug mode (-x)

    -v|--version                Print version.

    -h|--help                   Print this help and exit.


## Installation

Simple copy file in $PATH and

    $ chmod +x sitemap.sh

## Requirement

This script requires this command: wget, sed, awk, grep, cut and sort.
Optional: tee (for verbose mode), id or whoami (for root user check).

## Warnings

**THIS IS ONLY A TESTING SCRIPT** to generate sitemap in my situation.

It was written quickly, so it has errors and *ugliness* of course.

It is here because I need a public place to keep it, but if you need a sitemap generator try one of this:
https://code.google.com/archive/p/sitemap-generators/wikis/SitemapGenerators.wiki

## Note

If you have advice and suggestions to give, you are welcome.

I'm sorry for my bad english


ChangeLog.md
= Changelog

== 1.5 ==

- Fixed whoami error

== 1.4 ==

- Move color output to posix
- Clean check extension files

== 1.3 ==

- Added color output

== 1.2 ==

- Added fonts to exclusion list
- Added privacy and cookie page to exclusion list

== 1.1 ==

- Added check of duplicate entries. Although in theory it was already present thanks to the "sort -u" command.
- Added check-ssl option to consider duplicate files with the same filename but http or https protocol.

== 1.0 ==

- Ported to posix shells.
  Checked with shellcheck and tested with bash, zsh, csh, ksh, tcsh, dash.
- Added variable for mktemp (you can choose to use tempfile for example).

== 0.6 ==

- Added verbose ouput option
- Added rejected extensions option
- Added debug option

== 0.5.1 ==

- Fixed error in xml tag introduced in the previous version. I'm sorry.

== 0.5 ==

- Set ipv4 as default.
  Then removed the relative option and left only the one for ipv6.
- Added sitemap filename option
- Some cosmetic changes

== 0.4 ==

- Changed name in sitemap-generator
- Added ipv4 and ipv6 parameter
- Combined wget and output parsing, because in some circumstances wget exit without giving error
- Added check for wget and parsing ouput

== 0.3 ==

- Removed getops to also use long options
- Added check frequency parameter
- Added check priority parameter
- Added version parameter
- Fixed double quote to prevent globbing and word splitting in according with shellcheck

== 0.2 ==

- Added -r to read in according with shellcheck.
- Removed -e from getops.
  IT needed for the "rejected" in wget, but then I decided not to use it
- Readded -i parameters in case statement.
  I had temporarily removed in development stage. I'm sorry.
- Some aesthetic improvement.
- Added err function, as recommended in https://google.github.io/styleguide/shell.xml


== 0.1 ==

- Initial project


sitemap.sh
#!/bin/bash
#
# Copyright 2019 Corrado Franco (http://conraid.net)
# All rights reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version, with the following exception:
# the text of the GPL license may be omitted.
#
# This program is distributed in the hope that it will be useful, but
# without any warranty; without even the implied warranty of
# merchantability or fitness for a particular purpose. Compiling,
# interpreting, executing or merely reading the text of the program
# may result in lapses of consciousness and/or very being, up to and
# including the end of all existence and the Universe as we know it.
# See the GNU General Public License for more details.
#
# You may have received a copy of the GNU General Public License along
# with this program (most likely, a file named COPYING).  If not, see
# <http://www.gnu.org/licenses/>.

# Thanks to barkeep for idea
# http://www.lostsaloon.com/technology/how-to-create-an-xml-sitemap-using-wget-and-shell-script/

# N.B.
# THIS IS ONLY A TESTING SCRIPT to generate sitemap in my situation.
#
# If you need a generic and/or better sitemap generator try one of this
# https://code.google.com/archive/p/sitemap-generators/wikis/SitemapGenerators.wiki

VERSION=0.3

# Exit on error and undeclared variables
set -eu

# THE DEFAULTS INITIALIZATION - OPTIONALS
DEFAULT_EXT="php,html"
DEFAULT_INDEX="index.php"
DEFAULT_PRIORITY="0.5"
DEFAULT_FREQ="weekly"

LISTFILE=$(mktemp) || { echo "Failed to create temp file"; exit 1; }
SORTFILE=$(mktemp) || { echo "Failed to create temp file"; exit 1; }

# Not Root. And not sudo please ;)
if [[ "$EUID" = 0 ]]; then
  err "This script should not be run as root" 1>&2
  exit 1
fi

# Show script description
function info() {
cat << EOF
  This script crawls a web site from a given starting local URL or
  remote URL and generates a Sitemap file in the format that is
  accepted by Google.
  It does not follow links to other web sites or parent directory.
  It also respects robots.txt file.
EOF
help
}

# Print error in STDERR
function err() {
  echo -e "\n  [$(date +'%Y-%m-%d %H:%M:%S %z')]:" "$@" "\n" >&2
  exit
}

# Show help usate
function  help() {
cat << EOF

  Usage:
  $(basename "$0") [-r|--remote <url>] [-l|--locale <url>] [-p|--priority <number>] [-f|--frequency <string>] [-i|--index <string>] [-d|--docroot <path>] [-a|--accepted <ext>] [-h|--help] [-v|--version]


  Example:
  $(basename "$0") -l https://localhost/foobar/ -r https://example.com -d /home/html/foobar -p 0.8 -f daily

  Options:
    -r|--remote <url>      : Remote domain

    -l|--local <url>       : Local domain (ex. http://localhost/foobar/
                             Not with filename (ex. http://localhost/foo/bar.php)

    -p|--priority <number> : Priority. Valid values range from 0.0 to 1.0.
                             Default is "0.5"

    -f|--frequency <string>: Frequency. Valid values are:
                             always, hourly, daily, weekly, monthly, yearly, never
                             Default is "weekly"

    -i|--index <string>    : Name of index file
                             Default is "index.php"

    -d|--docroot <path>    : Doc Root

    -a|--accepted <ext>    : Comma-separated list of accepted extensions.
                             Default is "php,html"

    -v|--version           : Print version

    -h|--help              : Print this help and exit

EOF
}

# Check priority parameter
function check_priority() {
  if echo "$1" | grep -q "^0.[1-9]$" || [ "$1" == '1.0' ] ; then
    PRIORITY_DEFAULT="$1"
  else
    err "Valid values for 'priority' range from 0.0 to 1.0."
  fi
}

# Check frequency parameter
function check_freq() {
  if [ "$1" == "always" ] || [ "$1" == "hourly" ]  || [ "$1" == "daily" ] || [ "$1" == "weekly" ] || [ "$1" == "monthly" ] || [ "$1" == "yearly" ] || [ "$1" == "never" ]; then
    FREQ_DEFAULT="$1"
  else
    err "Valid values for 'frequency' are always, hourly, daily, weekly, monthly, yearly or never."
  fi
}

# Read parameters
while test $# -gt 0; do
  case "$1" in
    -r|--remote)
      test $# -lt 2 && err "Missing value for the optional argument '$1'." || REMOTEURL="${2%/}"
      shift
      ;;
    --remote=*)
      REMOTEURL="${1##--remote=}"
      ;;
    -r*)
      REMOTEURL="${1##-r}"
      ;;
    -l|--local)
      test $# -lt 2 && err "Missing value for the optional argument '$1'." || LOCALURL="$2"
      shift
      ;;
    --local=*)
      LOCALURL="${1##--locale=}"
      ;;
    -l*)
      LOCALURL="${1##-l}"
      ;;
    -p|--priority)
      test $# -lt 2 && err "Missing value for the optional argument '$1'." || check_priority "$2"
      shift
      ;;
    --priority=*)
      check_priority "${1##--priority=}"
      ;;
    -p*)
      check_priority "${1##-p}"
      ;;
    -f|--frequency)
      test $# -lt 2 && err "Missing value for the optional argument '$1'." || check_freq "$2"
      shift
      ;;
    --frequency=*)
       check_freq "${1##--frequency=}"
      ;;
    -f*)
      check_freq "${1##-f}"
      ;;
    -i|--index)
      test $# -lt 2 && err "Missing value for the optional argument '$1'." || INDEX_FILE="$2"
      shift
      ;;
    --index=*)
      INDEX_FILE="${1##--index=}"
      ;;
    -i*)
      INDEX_FILE="${1#-i}"
      ;;
    -d|--docroot)
      test $# -lt 2 && err "Missing value for the optional argument '$1'." || DOCROOT="$2"
      shift
      ;;
    --docroot=*)
      DOCROOT="${1##--docroot=}"
      ;;
    -d*)
      DOCROOT="${1##-d}"
      ;;
    -a|--accepted)
      test $# -lt 2 && err "Missing value for the optional argument '$1'." || ACCEPTED_EXT="$2"
      ACCEPTED_EXT="$2"
      shift
      ;;
    --accepted=*)
      ACCEPTED_EXT="${1##--accepted=}"
      ;;
    -a*)
      ACCEPTED_EXT="${1##-a}"
      ;;
    -h|--help)
      help
      exit 0
      ;;
    -h*)
      help
      exit 0
      ;;
    -v|--version)
      echo Version: "$VERSION"
      exit 0
      ;;
    -v*)
      echo Version: "$VERSION"
      exit 0
      ;;
    *)
      err "FATAL ERROR: Got an unexpected argument '$1'"
      ;;
  esac
  shift
done

# Check accepted extensions
if [[ -z ${ACCEPTED_EXT:-""} ]]; then
  ACCEPTED_EXT="${ACCEPTED_EXT:-$DEFAULT_EXT}"
elif [ "${ACCEPTED_EXT: -1}" == "," ] ; then
  ACCEPTED_EXT="${ACCEPTED_EXT::-1}"
else
  ACCEPTED_EXT="${ACCEPTED_EXT//[[:space:]]/}"
fi

# Check if at least one of A and B is present.
# If so, set the URLSCAN variable based on the parameters passed.
if [ -z "${LOCALURL:-""}" ] && [ -z "${REMOTEURL:-""}" ]; then
  err "$(basename "$0") requires -r or -l."
  help
  exit 1
elif [ -z "${LOCALURL:-""}" ]; then
  URLSCAN="$REMOTEURL"
elif [ "${LOCALURL: -1}" != "/" ] ; then
  URLSCAN="${LOCALURL}/"
else
  URLSCAN="${LOCALURL}"
fi

# Set default value of the variables
indexfile="${INDEX_FILE:-$DEFAULT_INDEX}"

# Make <url>
function makeurl() {

  priority="${PRIORITY_DEFAULT:-$DEFAULT_PRIORITY}"
  freq="${FREQ_DEFAULT:-$DEFAULT_FREQ}"

  if [ -n "${DOCROOT:-""}" ]; then
    FILE=$(echo "$1" | sed 's|/$||' | sed "s|${LOCALURL%/}|$DOCROOT|")
    FILENAME=$(basename "$FILE")

    if [ -d "$FILE" ]; then
      FILENAME="$indexfile"
      FILE="${FILE}"/"$FILENAME"
    fi
    if [ -f "$FILE" ]; then
      lastmod=$(date -r "$FILE" +%F)
    else
      # Show a error but not exit.
      err "FILE $FILE not exists. Check parameters"
    fi
  fi

  if [ -z "${LOCALURL:-""}" ] || [ -z "${REMOTEURL:-""}" ]; then
    remotefile="$1"
  else
    remotefile=${1/$URLSCAN/$REMOTEURL/}
  fi

  # This work for me.
  # Begin
  if [[ $1 = */ ]]; then
    priority="1"
    freq=daily
  elif [[ $1 = *legal* ]]; then
    priority="0.1"
    freq=monthly
  elif [[ $1 = *privacy* ]]; then
    priority="0.1"
    freq=monthly
  elif [[ $1 = *cookie* ]]; then
    priority="0.1"
    freq=monthly
  fi
  # End


  echo "Add $remotefile"
  {
    echo "<url>"
    echo "  <loc>$remotefile</loc>"
    [[ -z ${DOCROOT:-""} ]] || echo "  <lastmod>$lastmod</lastmod>"
    echo "  <changefreq>$freq</changefreq>"
    echo "  <priority>$priority</priority>"
    echo "</url>"
  } >> sitemap.xml
}

# Crawler
echo "Scan $URLSCAN"
wget --spider -r -nd -l inf --no-verbose --no-check-certificate --output-file="$LISTFILE" -np -A "$ACCEPTED_EXT" "$URLSCAN"
grep -i URL "$LISTFILE" | awk -F 'URL:' '{print $2}' | cut -d" " -f1 | sort -u > "$SORTFILE"

# Create XML file
cat << EOF > sitemap.xml
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns:xhtml="http://www.w3.org/1999/xhtml"
        xsi:schemaLocation="
            http://www.sitemaps.org/schemas/sitemap/0.9
            http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
EOF

# Modify ACCEPTED_EXT for case statement
ACC_EXT="*/"
for i in $(echo "$ACCEPTED_EXT" | tr , " "); do
  ACC_EXT="$ACC_EXT |*.$i"
done

# Read SORTFILE and call makeurl
while read -r filelist; do
  eval "case \"$filelist\" in
	  robots.txt )
            ;;
          $ACC_EXT)
            makeurl \"$filelist\"
            ;;
          *)
            ;;
        esac"
done < "$SORTFILE"

# Close xml file
echo "</urlset>" >> sitemap.xml

# Compress sitemap.xml
gzip -9 -f sitemap.xml -c > sitemap.xml.gz || { err "Failed to zipper sitemap"; exit 1; }

# Delete temporary files
rm "$LISTFILE" "$SORTFILE" || { err "Failed to remove temporary files"; exit 1; }


sitemap-generator.sh
#!/bin/sh
#
# Copyright 2019-2020 Corrado Franco (https://corradofranco.it)
# All rights reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version, with the following exception:
# the text of the GPL license may be omitted.
#
# This program is distributed in the hope that it will be useful, but
# without any warranty; without even the implied warranty of
# merchantability or fitness for a particular purpose. Compiling,
# interpreting, executing or merely reading the text of the program
# may result in lapses of consciousness and/or very being, up to and
# including the end of all existence and the Universe as we know it.
# See the GNU General Public License for more details.
#
# You may have received a copy of the GNU General Public License along
# with this program (most likely, a file named COPYING).  If not, see
# <http://www.gnu.org/licenses/>.

# Thanks to barkeep for idea
# http://www.lostsaloon.com/technology/how-to-create-an-xml-sitemap-using-wget-and-shell-script/

# N.B.
# THIS IS ONLY A TESTING SCRIPT to generate sitemap in my situation.
#
# If you need a generic and/or better sitemap generator try one of this
# https://code.google.com/archive/p/sitemap-generators/wikis/SitemapGenerators.wiki

# Exit on error and undeclared variables
set -eu

### VARIABLES ###

# Set script version
VERSION=1.5

# THE DEFAULT INITIALIZATIONS - OPTIONALS
DEFAULT_INDEX="index.php"
DEFAULT_PRIORITY="0.5"
DEFAULT_FREQ="weekly"
DEFAULT_OUTPUT="sitemap.xml"

# Set the bin utility program
MKTEMP="/usr/bin/mktemp" # Path of mktemp, or another temporary file creator (ex. tempfile).

# Set the default wget options
WGET_OPTIONS="--spider -r -nd -l inf --no-verbose --no-check-certificate -np"

# Color variables
RED_BEGIN="\\033[1;31m"
RED_END="\\033[0;39m"
GREEN_BEGIN="\\033[1;32m"
GREEN_END="\\033[0;39m"

### FUNCTIONS ###

# Print error in STDERR
err() {
  printf "\n [$(date +'%Y-%m-%d %H:%M:%S %z')]: %b %s %b \n\n" "$RED_BEGIN" "$*" "$RED_END" >&2
  exit 1
}

# Check root user
check_root() {
if [ -x "$(command -v id)" ]; then
  if test "$(id -u)" = "0"; then
    echo "This script should not be run as root"
    exit 1
  fi
elif [ -x "$(command -v whoami)" ]; then
  if test "$(whoami)" = "root"; then
    echo "This script should not be run as root"
    exit 1
  fi
else
  echo "Unable to determine if the script is run with the root user, consider whether to continue or not."
  read -r -p "Do you want continue? [y/N]" CONTINUE
  if [ "${CONTINUE}" = "y" ] || [ "${CONTINUE}" = "Y" ]; then
	echo "Ok, continue"
	sleep 5
  else
	exit 1
  fi
fi
}

# Show script description
info() {
cat << EOF
  This script crawls a web site from a given starting local URL or
  remote URL and generates a Sitemap file in the format that is
  accepted by Google.
  It does not follow links to other web sites or parent directory.
  It also respects robots.txt file.
EOF
help
}

check_program() {
  if ! [ -x "$(command -v "$1")" ]; then
      err "'$(basename "$1")' it's not present in path"
  fi
}

# Show help usate
 help() {
cat << EOF

Usage:
  $(basename "$0") [-r|--remote <url>] [-l|--locale <url>] [-p|--priority <number>] [-f|--frequency <string>] [-i|--index <string>] [-d|--docroot <path>] [-A|--accept <list>] [-R|-reject <list>] [-o|--output-file] [-6] [-ssl|--check-ssl] [-h|--help] [-V|--version] [-vv|--verbose] [--debug]


Example:
  $(basename "$0") -l https://localhost/foobar/ -r https://example.com -d /home/html/foobar -p 0.8 -f daily

Options:
 -r|--remote <url>           Set the remote URL.

 -l|--local <url>            Set the local URL (ex. http://localhost/foobar/ )
                             Not with filename (ex. http://localhost/foo/bar.php)

 -p|--priority <value>       Set the priority. Valid values range from 0.0 to 1.0.
                             Default is "0.5".

 -f|--frequency <value>      Set the frequency. Valid values are:
                             always, hourly, daily, weekly, monthly, yearly, never
                             Default is "weekly".

 -i|--index <filename>       Set the name of index file.
                             The default filename is "index.php".

 -d|--docroot <path>         Set dhe "Doc Root".

 -A|--accept <list>          Comma-separated list of accepted extensions.
                             Default is all.

 -R|--reject <list>          Comma-separated list of rejected extensions.
                             Default is nothing.

 -o|--output-file <filename> Set the name of the geneated sitemap file.
                             The default file name is sitemap.xml.

 -6                          Set the inet6-only to wget.
                             Connect only to IPv6 addresses.

 -ssl|--check-ssl            Check if there are duplicate URLs with http and https.
                             Useful when there is a redirect "http to https" in .htaccess for example.

 -vv|--verbose               Print details when crawling with wget.

 --debug                     Set bash to debug mode (-x)

 -v|--version                Print version.

 -h|--help                   Print this help and exit.

EOF
}

# Check priority parameter
check_priority() {
  if echo "$1" | grep -q "^0.[1-9]$" || [ "$1" = '1.0' ] ; then
    PRIORITY_DEFAULT="$1"
  else
    err "Valid values for 'priority' range from 0.0 to 1.0."
  fi
}

# Check frequency parameter
check_freq() {
  if [ "$1" = "always" ] || [ "$1" = "hourly" ]  || [ "$1" = "daily" ] || [ "$1" = "weekly" ] || [ "$1" = "monthly" ] || [ "$1" = "yearly" ] || [ "$1" = "never" ]; then
    FREQ_DEFAULT="$1"
  else
    err "Valid values for 'frequency' are always, hourly, daily, weekly, monthly, yearly or never."
  fi
}


# Make <url>
makeurl() {

  # Remove protocol is present check_ssl option.
  if [ "${CHECK_SSL:-""}" = '1' ]; then
    URL=$(echo "$1" | sed 's|.*://||')
  else
    URL="$1"
  fi

  # Check if $URL it's present in sitemap xml file.
  if ! grep -Fq "$URL</loc>" "$OUTPUT_FILE"; then

    PRIORITY="${PRIORITY_DEFAULT:-$DEFAULT_PRIORITY}"
    FREQ="${FREQ_DEFAULT:-$DEFAULT_FREQ}"

    if [ -n "${DOCROOT:-""}" ]; then
      if [ "$(echo "$DOCROOT" | awk '{print substr($0,length,1)}')" != "/" ] ; then
	      DOCROOT="${DOCROOT}/"
      fi
      FILE=$(echo "$1" | sed "s|${LOCALURL}|$DOCROOT|")

      if [ -d "$FILE" ]; then
        FILE="${FILE}"/"${INDEX_FILE:-$DEFAULT_INDEX}"
      fi

      if [ -f "$FILE" ]; then
        LASTMOD=$(date -r "$FILE" +%F)
      else
        # Show a error but not exit.
        LASTMOD=""
        printf "\n %b FILE %s not exists %b Check parameters. \n\n" "$RED_BEGIN" "$FILE" "$RED_END"
      fi
    fi

    if [ -z "${LOCALURL:-""}" ] || [ -z "${REMOTEURL:-""}" ]; then
      REMOTEFILE="$1"
    else
      REMOTEFILE=$(echo "$1" | sed "s|$URLSCAN|$REMOTEURL/|")
    fi

    # This works for me.
    # If you don't need this feature and you want to leave only the default values, comment this part.
    # Begin
    case "$1" in
      */ )
        PRIORITY="1"
        FREQ=daily
        ;;
      *legal*|*privacy*|*cookie* )
        PRIORITY="0.2"
        FREQ=monthly
        ;;
    esac
    # End


    printf "\n %b Add %s %b \n\n" "$GREEN_BEGIN" "$REMOTEFILE" "$GREEN_END"
    {
      echo "<url>"
      echo "  <loc>$REMOTEFILE</loc>"
      if [ "${LASTMOD:-""}" ]; then echo "  <lastmod>$LASTMOD</lastmod>"; fi
      echo "  <changefreq>$FREQ</changefreq>"
      echo "  <priority>$PRIORITY</priority>"
      echo "</url>"
    } >> "$OUTPUT_FILE"

  fi
}

### MAIN PROGRAM ###

# Not Root. And not sudo please ;)
check_root

# Check id exists utility programs
check_program sed
check_program awk
check_program wget
check_program grep
check_program cut
check_program sort
check_program $MKTEMP

# Read parameters
while test $# -gt 0; do
  case "$1" in
    -r|--remote)
      test $# -lt 2 && err "Missing value for the optional argument '$1'."
      REMOTEURL="${2%/}"
      shift
      ;;
    --remote=*)
      REMOTEURL="${1##--remote=}"
      ;;
    -r*)
      REMOTEURL="${1##-r}"
      ;;
    -l|--local)
      test $# -lt 2 && err "Missing value for the optional argument '$1'."
      LOCALURL="$2"
      shift
      ;;
    --local=*)
      LOCALURL="${1##--local=}"
      ;;
    -l*)
      LOCALURL="${1##-l}"
      ;;
    -p|--priority)
      test $# -lt 2 && err "Missing value for the optional argument '$1'."
      check_priority "$2"
      shift
      ;;
    --priority=*)
      check_priority "${1##--priority=}"
      ;;
    -p*)
      check_priority "${1##-p}"
      ;;
    -f|--frequency)
      test $# -lt 2 && err "Missing value for the optional argument '$1'."
      check_freq "$2"
      shift
      ;;
    --frequency=*)
      check_freq "${1##--frequency=}"
      ;;
    -f*)
      check_freq "${1##-f}"
      ;;
    -i|--index)
      test $# -lt 2 && err "Missing value for the optional argument '$1'."
      INDEX_FILE="$2"
      shift
      ;;
    --index=*)
      INDEX_FILE="${1##--index=}"
      ;;
    -i*)
      INDEX_FILE="${1#-i}"
      ;;
    -o|--output-file)
      test $# -lt 2 && err "Missing value for the optional argument '$1'."
      OUTPUT_FILE="$2"
      shift
      ;;
    --output-file=*)
      OUTPUT_FILE="${1##--output=}"
      ;;
    -o*)
      OUTPUT_FILE="${1#-o}"
      ;;
    -d|--docroot)
      test $# -lt 2 && err "Missing value for the optional argument '$1'."
      DOCROOT="$2"
      shift
      ;;
    --docroot=*)
      DOCROOT="${1##--docroot=}"
      ;;
    -d*)
      DOCROOT="${1##-d}"
      ;;
    -A|--accept)
      test $# -lt 2 && err "Missing value for the optional argument '$1'."
      ACCEPT_EXT="$2"
      shift
      ;;
    --accept=*)
      ACCEPT_EXT="${1##--accept=}"
      ;;
    -A*)
      ACCEPT_EXT="${1##-a}"
      ;;
    -R|--reject)
      test $# -lt 2 && err "Missing value for the optional argument '$1'."
      REJECT_EXT="$2"
      shift
      ;;
    --reject=*)
      REJECT_EXT="${1##--reject=}"
      ;;
    -R*)
      REJECT_EXT="${1##-e}"
      ;;
    -6)
      WGET_OPTIONS="$WGET_OPTIONS -6"
      ;;
    -ssl|--check-ssl)
      CHECK_SSL=1
      ;;
    -h|--help)
      help
      exit 0
      ;;
    -h*)
      help
      exit 0
      ;;
    -V|--version)
      echo Version: "$VERSION"
      exit 0
      ;;
    -vv|--verbose)
      VERBOSE=1
      ;;
    --debug)
      set -x
      DEBUG=1
      ;;
    *)
      err "FATAL ERROR: Got an unexpected argument '$1'. Use -h for help."
      ;;
  esac
  shift
done

# Check if at least one of A and B is present.
# If so, set the URLSCAN variable based on the parameters passed.
if [ -z "${LOCALURL:-""}" ] && [ -z "${REMOTEURL:-""}" ]; then
  err "$(basename "$0") requires -r or -l. Use -h for help"
elif [ -z "${LOCALURL:-""}" ]; then
  URLSCAN="$REMOTEURL"
elif [ "$(echo "$LOCALURL" | awk '{print substr($0,length,1)}')" != "/" ] ; then
  LOCALURL="${LOCALURL}/"
  URLSCAN="${LOCALURL}"
else
  URLSCAN="${LOCALURL}"
fi

# Check accepted extensions
if [ -n "${ACCEPT_EXT:-""}" ]; then
  if [ "$(echo "$ACCEPT_EXT" | awk '{print substr($0,length,1)}')" = "," ] ; then
    WGET_OPTIONS="$WGET_OPTIONS -A $(echo "$ACCEPT_EXT" | sed 's/,$//')"
  else
    WGET_OPTIONS="$WGET_OPTIONS -A $(echo "$ACCEPT_EXT" | tr -d " ")"
  fi
fi


# Check rejected extensions
if [ -n "${REJECT_EXT:-""}" ]; then
  if [ "$(echo "$REJECT_EXT" | awk '{print substr($0,length,1)}')" = "," ] ; then
    WGET_OPTIONS="$WGET_OPTIONS -R $(echo "$REJECT_EXT" | sed 's/,$//')"
  else
    WGET_OPTIONS="$WGET_OPTIONS -R $(echo "$REJECT_EXT" | tr -d " ")"
  fi
fi

# Set files
OUTPUT_FILE="${OUTPUT_FILE:-$DEFAULT_OUTPUT}"
LISTFILE=$($MKTEMP --suffix=-list) || { err "Failed to create LIST temp file. $MKTEMP exists and/or /tmp is writable?"; }
SORTFILE=$($MKTEMP --suffix=-sort) || { err "Failed to create SORT temp file. $MKTEMP exists and/or /tmp is writable?"; }

# Crawler
printf "\n Scan: %b %s %b \n\n" "$GREEN_BEGIN" "$URLSCAN" "$GREEN_END"

if [ ${VERBOSE:-""} = '1' ]; then
  if ! [ -x "$(command -v tee)" ]; then
    err "Verbose mode is possible only with command 'tee' in path"
  fi
  # shellcheck disable=SC2086
  wget $WGET_OPTIONS "$URLSCAN" 2>&1 | tee "$LISTFILE"
else
  # shellcheck disable=SC2086
  wget $WGET_OPTIONS "$URLSCAN" -o "$LISTFILE" || true
fi

# Check if wget creates listfile
if ! [ -s "$LISTFILE" ]; then
  err "Error in wget. Check parameters, network, web server or other."
fi

# Select only URL line
grep -i URL "$LISTFILE" | sed 's/URL: /URL:/' | awk -F 'URL:' '{print $2}' | cut -d" " -f1 | sed '/^$/d' | sort -u > "$SORTFILE"

# Create XML file
cat << EOF > "$OUTPUT_FILE"
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns:xhtml="http://www.w3.org/1999/xhtml"
        xsi:schemaLocation="
            http://www.sitemaps.org/schemas/sitemap/0.9
            http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
EOF

# Read SORTFILE and call makeurl
while read -r FILELIST; do
  case "$FILELIST" in
    *robots*|*\.txt )
      ;;
    *privacy*|*cookie*)
      #makeurl "$FILELIST" # This works for me, if you need to insert this type of file too, uncomment the line.
      ;;
    *\.jpg|*\.gif|*\.jpeg|*\.ico|*\.png|*\.svg|*\.webp )
      #makeurl "$FILELIST" # This works for me, if you need to insert this type of file too, uncomment the line.
      ;;
    *\.jpg\?*|*\.gif\?*|*\.jpeg\?*|*\.ico\?*|*\.png\?*|*\.svg\?*|*\.webp\?* )
      #makeurl "$FILELIST" # This works for me, if you need to insert this type of file too, uncomment the line.
      ;;
    *\.eot|*\.ttf|*\.woff|*\.woff2|*\.otf )
      #makeurl "$FILELIST" # This works for me, if you need to insert this type of file too, uncomment the line.
      ;;
    *\.eot\?*|*\.ttf\?*|*\.woff\?*|*\.woff2\?*|*\.otf\?* )
      #makeurl "$FILELIST" # This works for me, if you need to insert this type of file too, uncomment the line.
      ;;
    *\.js|*\.css)
      #makeurl "$FILELIST" # This works for me, if you need to insert this type of file too, uncomment the line.
      ;;
    *\.js\?*|*\.css\?*)
      #makeurl "$FILELIST" # This works for me, if you need to insert this type of file too, uncomment the line.
      ;;
    *)
      makeurl "$FILELIST"
      ;;
  esac
done < "$SORTFILE"



# Close xml file
echo "</urlset>" >> "$OUTPUT_FILE"

# Compress output file (default sitemap.xml)
gzip -9 -f "$OUTPUT_FILE" -c > "${OUTPUT_FILE}".gz || { err "Failed to zipper sitemap"; }

# Delete temporary files
if [ -z ${DEBUG:-""} ]; then
  rm "$SORTFILE" "$LISTFILE" || { err "Failed to remove temporary files"; }
fi
